File: src/main/java/_core/BaseScreen.java
package _core;

import java.time.Duration;

import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import io.appium.java_client.AppiumDriver;
import io.appium.java_client.MobileElement;
import io.appium.java_client.pagefactory.AppiumFieldDecorator;

public class BaseScreen extends Capabilities {
	
	private AppiumDriver<MobileElement> driver = inicializarAppiumDriver();

	public BaseScreen () {
		
		PageFactory.initElements(new AppiumFieldDecorator(driver, Duration.ofSeconds(20)), this);
	}

	WebDriverWait wait = new WebDriverWait(driver, 5);
	
	public void click (MobileElement element) {
		
		try {
			wait.until(ExpectedConditions.elementToBeClickable(element)).click();			
		} catch (Exception e) {
			wait.until(ExpectedConditions.visibilityOf(element)).click();
		}
	}
	
	public void sendKeys (MobileElement element, String text) {
		
		try {
			wait.until(ExpectedConditions.elementToBeClickable(element)).sendKeys(text);			
		} catch (Exception e) {
			wait.until(ExpectedConditions.visibilityOf(element)).sendKeys(text);
		}
	}

}

File: src/main/java/qa/reinaldo/_core/BaseScreen.kt
package _core

import io.appium.java_client.AppiumDriver
import io.appium.java_client.MobileElement
import io.appium.java_client.pagefactory.AppiumFieldDecorator
import org.openqa.selenium.By
import org.openqa.selenium.support.PageFactory
import org.openqa.selenium.support.ui.ExpectedConditions
import org.openqa.selenium.support.ui.WebDriverWait
import java.time.Duration

open class BaseScreen : Capabilities() {

    private val driver: AppiumDriver<MobileElement>? = Capabilities.Companion.inicializarAppiumDriver()
    open var wait = WebDriverWait(driver, 30)

    fun click(element: MobileElement?) {
        try {
            wait.until(ExpectedConditions.elementToBeClickable(element)).click()
        } catch (e: Exception) {
            wait.until(ExpectedConditions.visibilityOf(element)).click()
        }
    }

    fun click(element: By) {
        try {
            wait.until(ExpectedConditions.elementToBeClickable(element)).click()
        } catch (e: Exception) {
            wait.until(ExpectedConditions.presenceOfElementLocated(element)).click()
        }
    }

    fun sendKeys(element: MobileElement?, text: String?) {
        try {
            wait.until(ExpectedConditions.elementToBeClickable(element)).sendKeys(text)
        } catch (e: Exception) {
            wait.until(ExpectedConditions.visibilityOf(element)).sendKeys(text)
        }
    }

    init {
        PageFactory.initElements(AppiumFieldDecorator(driver, Duration.ofSeconds(30)), this)
    }
}